lib/tree.rb

```diff
@@ -1,64 +1,155 @@
 class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
-
-   def initialize(key, val)
+  
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
   def initialize
     @root = nil
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def add_helper(current_node, key, value)
+    return TreeNode.new(key,value) if current_node.nil?
+    
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value) 
+    end
+    
+    return current_node
+  end
+  
+  # Time Complexity: O(log n)
+  # Space Complexity: O(log n)
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key,value)
+    
+    if @root.nil?
+      @root = new_node
+    else  
+      add_helper(@root, key, value)
+    end
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: Best: O(log n), Worst: O(n)
+  # Space Complexity: O(n)
```

Comment:
Both space/time complexity are O(log n) for a balanced tree and O(n) for an unbalanced tree.